File: 2/solution.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        added = 0
        head = tmp = ListNode(-1)
        while l1 or l2:
            val = added
            if l1:
                val += l1.val
                l1 = l1.next
            if l2:
                val += l2.val
                l2 = l2.next
            tmp.next = ListNode(val % 10)
            tmp = tmp.next
            added = val // 10
        if added:
            tmp.next = ListNode(added)
        return head.next

File: 2/solution.go
package _2

type ListNode struct {
	Val  int
	Next *ListNode
}

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	added := 0
	head := &ListNode{-1, nil}
	tmp := head
	for l1 != nil || l2 != nil {
		val := added
		if l1 != nil {
			val += l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			val += l2.Val
			l2 = l2.Next
		}
		tmp.Next = &ListNode{val % 10, nil}
		tmp = tmp.Next
		added = val / 10
	}
	if added != 0 {
		tmp.Next = &ListNode{added, nil}
	}
	return head.Next
}
